lib/parse/dep-graph.ts

```diff
@@ -0,0 +1,55 @@
+import type { MavenGraph, MavenGraphNode } from './types';
+
+import { DepGraph, DepGraphBuilder, PkgInfo } from '@snyk/dep-graph';
+import { parseDependency } from './dependency';
+
+export function buildDepGraph(mavenGraph: MavenGraph): DepGraph {
+  const { rootId, nodes } = mavenGraph;
+  const root = getPkgInfo(rootId);
+  const builder = new DepGraphBuilder({ name: 'maven' }, root);
+  const visited: string[] = [];
+  const queue: QueueItem[] = [];
+  queue.push(...getItems(rootId, nodes[rootId]));
+
+  // breadth first search
+  while (queue.length > 0) {
+    const item = queue.shift();
+    if (!item) continue;
+    const { id, parentId } = item;
+    const pkgInfo = getPkgInfo(id);
+    if (visited.includes(id)) {
+      const prunedId = id + ':pruned';
+      builder.addPkgNode(pkgInfo, prunedId, { labels: { pruned: 'true' } });
+      builder.connectDep(parentId, prunedId);
+      continue; // don't queue any more children
+    }
+    const parentNodeId = parentId === rootId ? builder.rootNodeId : parentId;
+    builder.addPkgNode(pkgInfo, id);
+    builder.connectDep(parentNodeId, id);
+    queue.push(...getItems(id, nodes[id]));
+    visited.push(id);
+  }
+
+  return builder.build();
+}
+
+interface QueueItem {
+  id: string;
+  parentId: string;
+}
+
+function getItems(parentId: string, node?: MavenGraphNode): QueueItem[] {
+  const items: QueueItem[] = [];
+  for (const id of node?.dependsOn || []) {
+    items.push({ id, parentId });
+  }
+  return items;
```

Comment:
If node doesn't depend on anything, then we have an empty array / a falsy value which we have to skip later ([L17](https://github.com/snyk/snyk-mvn-plugin/pull/135/files#diff-1942f2c6c5731c125ebd347cb9d04093ccce2f8206f568c4387da3dfc743b021R17)) - should we eliminate it earlier, e.g. not call getItems if there is no node with the ID we're looking at?

Reply:
`if (!item) continue;` is simply there to convince Typescript we're not at the end of the queue when doing `queue.shift()`, even though that will never happen because we're in a while loop that stops when the queue length is empty.

side note, empty arrays are not falsey, they're truthy - have a go in the node repl

```shell
node
Welcome to Node.js v14.17.6.
Type ".help" for more information.
> var arr = [];
undefined
> !!arr
true
> 
```